File: Cpp/BaekJoon/Sliver/S1_1743.cpp
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

struct Node {
  int y;
  int x;
};

int bfs(int n, int m, vector<vector<int>> map) {
  vector<vector<bool>> visited;
  visited.resize(n + 1, vector<bool>(m + 1, false));

  int dy[] = {-1, 1, 0, 0};
  int dx[] = {0, 0, -1, 1};

  int result = 0;

  for (int i = 1; i <= n; i++) {
    for (int j = 1; j <= m; j++) {
      if (map[i][j] == 1 && !visited[i][j]) {
        queue<Node> que;

        que.push({i, j});
        visited[i][j] = true;

        int cnt = 1;

        while (!que.empty()) {
          Node now = que.front();
          que.pop();

          for (int i = 0; i < 4; i++) {
            int ny = now.y + dy[i];
            int nx = now.x + dx[i];

            if (ny < 1 || nx < 1 || ny > n || nx > m) continue;
            if (map[ny][nx] == 0) continue;
            if (visited[ny][nx]) continue;

            que.push({ny, nx});
            visited[ny][nx] = true;

            cnt++;
          }
        }

        if (result < cnt) result = cnt;
      }
    }
  }

  return result;
}

int main() {
  int n, m, k;
  cin >> n >> m >> k;

  vector<vector<int>> map;
  map.resize(n + 1, vector<int>(m + 1, 0));

  for (int i = 0; i < k; i++) {
    int row, col;
    cin >> row >> col;

    map[row][col] = 1;
  }

  int result = bfs(n, m, map);

  cout << result;

  return 0;
}

File: Python/BaekJoon/Sliver/S1_1743.py
from collections import deque

n, m, k = map(int, input().split())

location = [[0 for _ in range(m + 1)] for _ in range(n + 1)]
visited = [[False for _ in range(m + 1)] for _ in range(n + 1)]

for i in range(k):
    row, col = map(int, input().split())

    location[row][col] = 1

dy = [-1, 1, 0, 0]
dx = [0, 0, -1, 1]

result = 0

for i in range(1, n + 1):
    for j in range(1, m + 1):

        if ((location[i][j] == 1) and (visited[i][j] == False)):
            que = deque()
            que.append((i, j))

            visited[i][j] = True

            cnt = 1

            while(len(que) != 0):
                now = que.popleft()

                for dir in range(4):
                    ny = now[0] + dy[dir]
                    nx = now[1] + dx[dir]

                    if (ny < 1 or nx < 1 or ny > n or nx > m): continue
                    if (location[ny][nx] != 1): continue
                    if (visited[ny][nx]): continue

                    que.append((ny, nx))
                    visited[ny][nx] = True
                    
                    cnt += 1
        
            if (result < cnt): result = cnt

print(result)